src/recordlinker/linkage/matchers.py

```diff
@@ -0,0 +1,427 @@
+
+    :param record_i: One of the records in the candidate pair to evaluate.
+    :param record_j: The second record in the candidate pair.
+    :param feature_col: The name of the column being evaluated (e.g. "city").
+    :param col_to_idx: A dictionary mapping column names to the numeric index
+      in which they occur in order in the data.
+    :return: A boolean indicating whether the features are an exact match.
+    """
+    idx = col_to_idx[feature_col]
+    return record_i[idx] == record_j[idx]
+
+
+def feature_match_four_char(
+    record_i: list,
+    record_j: list,
+    feature_col: str,
+    col_to_idx: dict[str, int],
+    **kwargs: dict,
+) -> bool:
+    """
+    Determines whether a string feature in a pair of records exactly matches
+    on the first four characters.
+
+    :param record_i: One of the records in the candidate pair to evaluate.
+    :param record_j: The second record in the candidate pair.
+    :param feature_col: The name of the column being evaluated (e.g. "city").
+    :param col_to_idx: A dictionary mapping column names to the numeric index
+      in which they occur in order in the data.
+    :return: A boolean indicating whether the features are a match.
+    """
+    idx = col_to_idx[feature_col]
+    first_four_i = record_i[idx][: min(4, len(record_i[idx]))]
+    first_four_j = record_j[idx][: min(4, len(record_j[idx]))]
+    return first_four_i == first_four_j
+
+
+def feature_match_fuzzy_string(
+    record_i: list,
+    record_j: list,
+    feature_col: str,
+    col_to_idx: dict[str, int],
+    **kwargs: dict,
+) -> bool:
+    """
+    Determines whether two strings in a given pair of records are close
+    enough to constitute a partial match. The exact nature of the match
+    is determined by the specified string comparison function (see
+    harmonization/utils/compare_strings for more details) as well as a
+    scoring threshold the comparison must meet or exceed.
+
+    :param record_i: One of the records in the candidate pair to evaluate.
+    :param record_j: The second record in the candidate pair.
+    :param feature_col: The name of the column being evaluated (e.g. "city").
+    :param col_to_idx: A dictionary mapping column names to the numeric index
+      in which they occur in order in the data.
+    :param **kwargs: Optionally, a dictionary including specifications for
+      the string comparison metric to use, as well as the cutoff score
+      beyond which to classify the strings as a partial match.
+    :return: A boolean indicating whether the features are a fuzzy match.
+    """
+    idx = col_to_idx[feature_col]
+
+    # Convert datetime obj to str using helper function
+    if feature_col == "birthdate":
+        record_i[idx] = utils.datetime_to_str(record_i[idx])
+        record_j[idx] = utils.datetime_to_str(record_j[idx])
+
+    # Special case for two empty strings, since we don't want vacuous
+    # equality (or in-) to penalize the score
```

Comment:
@ericbuckley This comment answers your previous question of: what happens when two fields are missing? How is equality evaluated?

Reply:
Well this is half of it, I'm also interested in how the blocking queries are meant to behave when PII is missing.